src/wallet/wallet.cpp

```diff
@@ -3987,37 +3991,54 @@ CWallet* CWallet::CreateWalletFromFile(const std::string& name, const fs::path&
             return nullptr;
         }
         walletInstance->SetMaxVersion(nMaxVersion);
-    }
 
-    if (fFirstRun)
-    {
         // ensure this wallet.dat can only be opened by clients supporting HD with chain split and expects no default key
-        if (!gArgs.GetBoolArg("-usehd", true)) {
+        if (!gArgs.GetBoolArg("-usehd", true) && fFirstRun) {
             InitError(strprintf(_("Error creating %s: You can't create non-HD wallets with this version."), walletFile));
             return nullptr;
         }
-        walletInstance->SetMinVersion(FEATURE_NO_DEFAULT_KEY);
 
-        // generate a new master key
-        CPubKey masterPubKey = walletInstance->GenerateNewHDMasterKey();
-        if (!walletInstance->SetHDMasterKey(masterPubKey))
-            throw std::runtime_error(std::string(__func__) + ": Storing master key failed");
+        // Upgrade to HD if necessary
+        {
+            LOCK(walletInstance->cs_wallet);
+            bool hd_upgrade = false;
+            if (walletInstance->CanSupportFeature(FEATURE_HD) && !walletInstance->IsHDEnabled()) {
+                LogPrintf("Upgrading wallet to HD\n");
+                walletInstance->SetMinVersion(FEATURE_HD);
+
+                // generate a new master key
+                CPubKey masterPubKey = walletInstance->GenerateNewHDMasterKey();
+                if (!walletInstance->SetHDMasterKey(masterPubKey))
+                    throw std::runtime_error(std::string(__func__) + ": Storing master key failed");
+                hd_upgrade = true;
+            }
+            // Upgrade to HD chain split if necessary
+            if (walletInstance->CanSupportFeature(FEATURE_HD_SPLIT)) {
+                LogPrintf("Upgrading wallet to use HD chain split\n");
+                walletInstance->SetMinVersion(FEATURE_HD_SPLIT);
```

Comment:
Same comment here, this implicitly will generate a new keypool on next internal key usage.

Reply:
Is that a problem though? An upgrade was explicitly asked for. Would it be better to regenerate the keypool in this instance too? Either way, both the internal and external keypools will still be generated from the same seed.

Reply:
No, this significantly changes the upgradewallet semantics, before you could upgrade and you'd still be able to use your backup in the same way based on the keypool size, now that is no longer true, which needs huge flashing warnings, at a minimum.

Reply:
Good point, @TheBlueMatt. I was forgetting about the distinction "need a new backup" and "need new software in the future".

What would you suggest as an ideal solution? `-upgradewallet` by default never making a change that requires a new backup, while having another option (`-fullupgradewallet` ?) to indicate you're aware a new backup is needed?

Reply:
What about an `InitWarning` message that appears when the keypool is regenerated that a new backup will need to be made?